File: src/mcp_server.py
"""
MCP Server for CV Database Access
Provides standardized tools for querying PostgreSQL and Qdrant vector DB
"""

import logging
from typing import Dict, List, Any, Optional
from qdrant_client import QdrantClient
from qdrant_client.models import Filter, FieldCondition, MatchValue
from langchain_openai import OpenAIEmbeddings

from config import get_config
from db_manager import get_db_manager, convert_dates

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


# ============================================================================
# HELPER FUNCTIONS
# ============================================================================

def _format_result(tool_name: str, status: str, results: Any = None, error: str = None, **extra_fields) -> Dict[str, Any]:
    """
    Format standardized response for all MCP tools.

    Args:
        tool_name: Name of the tool that was called
        status: "success" or "error"
        results: Tool results (list or dict)
        error: Error message if status is "error"
        **extra_fields: Additional fields to include (e.g., company, category, date_range)

    Returns:
        Formatted result dictionary
    """
    response = {
        "status": status,
        "tool": tool_name,
    }

    if status == "success":
        if results is not None:
            if isinstance(results, list):
                response["results_count"] = len(results)
                response["results"] = results
            elif isinstance(results, dict):
                response.update(results)
            else:
                response["results"] = results
    else:
        response["error"] = error

    # Add any extra fields (like company, category, date_range, etc.)
    response.update(extra_fields)

    return response


# ============================================================================
# DATABASE TOOLS
# ============================================================================

class DatabaseTools:
    """MCP Tools for accessing CV database"""

    def __init__(self, config=None):
        """Initialize DatabaseTools with configuration"""
        self.config = config or get_config()
        self.db = get_db_manager(self.config)
        self.embedding_model = OpenAIEmbeddings(
            api_key=self.config.get_openai_api_key(),
            model=self.config.get_embedding_model()
        )
        self.qdrant_client = QdrantClient(
            url=self.config.get_qdrant_url(),
            api_key=self.config.get_qdrant_api_key()
        )
        # Cache CV ID to avoid repeated lookups
        self._cv_id_cache: Optional[str] = None
        logger.info("DatabaseTools initialized")

    def _get_cv_id(self) -> str:
        """
        Get the CV ID from the database with caching

        Returns:
            str: CV ID or "default-cv-id" if not found
        """
        # Return cached value if available
        if self._cv_id_cache:
            return self._cv_id_cache

        try:
            results = self.db.execute_query(
                "SELECT DISTINCT cv_id FROM skills LIMIT 1",
                as_dict=False
            )

            if results:
                self._cv_id_cache = str(results[0][0])
                return self._cv_id_cache
            else:
                logger.warning("No CV ID found in database, using default")
                return "default-cv-id"
        except Exception as e:
            logger.warning(f"Could not fetch CV ID: {e}, using default")
            return "default-cv-id"

    # ========================================================================
    # TOOL 1: Get CV Summary
    # ========================================================================
    def get_cv_summary(self) -> Dict[str, Any]:
        """
        Get a summary of the CV including name, role, experience, and key stats.

        Returns:
            Dict with CV summary information
        """
        try:
            results = self.db.execute_query(
                """
                SELECT name, crole as current_role, total_years_experience,
                       total_jobs, total_degrees, total_publications,
                       domains, all_skills
                FROM cv_summary
                LIMIT 1
                """
            )

            if results:
                summary_dict = results[0]
                logger.info("CV summary retrieved successfully")
                return _format_result("get_cv_summary", "success", summary=summary_dict)
            else:
                logger.warning("CV not found in database")
                return _format_result("get_cv_summary", "error", error="CV not found")

        except Exception as e:
            logger.error(f"Error in get_cv_summary: {e}")
            return _format_result("get_cv_summary", "error", error=str(e))

    # ========================================================================
    # TOOL 2: Search Company Experience
    # ========================================================================
    def search_company_experience(self, company_name: str) -> Dict[str, Any]:
        """
        Find all jobs at a specific company.

        Args:
            company_name: Name of the company (e.g., 'KasiOss')

        Returns:
            Dict with work experience records
        """
        try:
            cv_id = self._get_cv_id()
            results = self.db.execute_query(
                """
                SELECT company, role, location, start_date, end_date, is_current,
                       technologies, skills, domain, seniority, team_size
                FROM work_experience
                WHERE cv_id = %s AND company ILIKE %s
                ORDER BY start_date DESC
                """,
                (cv_id, f"%{company_name}%")
            )

            # Convert dates to strings
            for result in results:
                convert_dates(result, ["start_date", "end_date"])

            logger.info(f"Found {len(results)} jobs at {company_name}")
            return _format_result(
                "search_company_experience",
                "success",
                results=results,
                company=company_name
            )

        except Exception as e:
            logger.error(f"Error in search_company_experience: {e}")
            return _format_result("search_company_experience", "error", error=str(e))

    # ========================================================================
    # TOOL 3: Search Technology Experience
    # ========================================================================
    def search_technology_experience(self, technology: str) -> Dict[str, Any]:
        """
        Find all jobs using a specific technology.

        Args:
            technology: Technology name (e.g., 'Python', 'TensorFlow')

        Returns:
            Dict with work experience using the technology
        """
        try:
            cv_id = self._get_cv_id()
            results = self.db.execute_query(
                """
                SELECT company, role, start_date, end_date, technologies, domain
                FROM work_experience
                WHERE cv_id = %s AND %s = ANY(technologies)
                ORDER BY start_date DESC
                """,
                (cv_id, technology)
            )

            # Convert dates
            for result in results:
                convert_dates(result, ["start_date", "end_date"])

            logger.info(f"Found {len(results)} jobs using {technology}")
            return _format_result(
                "search_technology_experience",
                "success",
                results=results,
                technology=technology
            )

        except Exception as e:
            logger.error(f"Error in search_technology_experience: {e}")
            return _format_result("search_technology_experience", "error", error=str(e))

    # ========================================================================
    # TOOL 4: Search Work by Date Range
    # ========================================================================
    def search_work_by_date(self, start_year: int, end_year: int) -> Dict[str, Any]:
        """
        Find work experience within a date range.

        Args:
            start_year: Start year (e.g., 2020)
            end_year: End year (e.g., 2024)

        Returns:
            Dict with work experience in the date range
        """
        try:
            cv_id = self._get_cv_id()
            results = self.db.execute_query(
                """
                SELECT company, role, start_date, end_date, technologies, keywords
                FROM work_experience
                WHERE cv_id = %s
                  AND start_date >= %s::date
                  AND (end_date <= %s::date OR end_date IS NULL)
                ORDER BY start_date DESC
                """,
                (cv_id, f"{start_year}-01-01", f"{end_year}-12-31")
            )

            for result in results:
                convert_dates(result, ["start_date", "end_date"])

            logger.info(f"Found {len(results)} jobs between {start_year}-{end_year}")
            return _format_result(
                "search_work_by_date",
                "success",
                results=results,
                date_range=f"{start_year}-{end_year}"
            )

        except Exception as e:
            logger.error(f"Error in search_work_by_date: {e}")
            return _format_result("search_work_by_date", "error", error=str(e))

    # ========================================================================
    # TOOL 5: Search Education
    # ========================================================================
    def search_education(self, institution: Optional[str] = None, degree: Optional[str] = None) -> Dict[str, Any]:
        """
        Find education records by institution or degree.

        Args:
            institution: University or institution name (optional)
            degree: Degree type (e.g., 'PhD', 'Master') (optional)

        Returns:
            Dict with education records
        """
        try:
            cv_id = self._get_cv_id()

            if institution:
                results = self.db.execute_query(
                    """
                    SELECT institution, degree, field, specialization, graduation_date, thesis, publications
                    FROM education
                    WHERE cv_id = %s AND institution ILIKE %s
                    """,
                    (cv_id, f"%{institution}%")
                )
                search_type = f"institution: {institution}"

            elif degree:
                results = self.db.execute_query(
                    """
                    SELECT institution, degree, field, specialization, graduation_date, thesis
                    FROM education
                    WHERE cv_id = %s AND degree ILIKE %s
                    """,
                    (cv_id, f"%{degree}%")
                )
                search_type = f"degree: {degree}"

            else:
                results = self.db.execute_query(
                    """
                    SELECT institution, degree, field, specialization, graduation_date, thesis
                    FROM education
                    WHERE cv_id = %s
                    """,
                    (cv_id,)
                )
                search_type = "all education"

            for result in results:
                convert_dates(result, ["graduation_date"])

            logger.info(f"Found {len(results)} education records for {search_type}")
            return _format_result(
                "search_education",
                "success",
                results=results,
                search_type=search_type
            )

        except Exception as e:
            logger.error(f"Error in search_education: {e}")
            return _format_result("search_education", "error", error=str(e))

    # ========================================================================
    # TOOL 6: Search Publications
    # ========================================================================
    def search_publications(self, year: Optional[int] = None) -> Dict[str, Any]:
        """
        Search publications by year.

        Args:
            year: Publication year (optional, defaults to all publications)

        Returns:
            Dict with publications
        """
        try:
            cv_id = self._get_cv_id()

            if year:
                results = self.db.execute_query(
                    """
                    SELECT title, year, conference_name, doi, keywords, content_text
                    FROM publications
                    WHERE cv_id = %s AND year = %s
                    ORDER BY year DESC
                    """,
                    (cv_id, year)
                )
                search_type = f"year: {year}"

            else:
                results = self.db.execute_query(
                    """
                    SELECT title, year, conference_name, doi, keywords, content_text
                    FROM publications
                    WHERE cv_id = %s
                    ORDER BY year DESC
                    """,
                    (cv_id,)
                )
                search_type = "all publications"

            logger.info(f"Found {len(results)} publications for {search_type}")
            return _format_result(
                "search_publications",
                "success",
                results=results,
                search_type=search_type
            )

        except Exception as e:
            logger.error(f"Error in search_publications: {e}")
            return _format_result("search_publications", "error", error=str(e))

    # ========================================================================
    # TOOL 7: Search Skills
    # ========================================================================
    def search_skills(self, category: str) -> Dict[str, Any]:
        """
        Find skills by category.

        Args:
            category: Skill category (e.g., 'AI', 'ML', 'programming', 'Tools', 'Cloud', 'Data_tools')

        Returns:
            Dict with skills in the category
        """
        try:
            cv_id = self._get_cv_id()
            results = self.db.execute_query(
                """
                SELECT skill_name
                FROM skills
                WHERE cv_id = %s AND skill_category = %s
                ORDER BY skill_name
                """,
                (cv_id, category)
            )

            logger.info(f"Found {len(results)} skills in category {category}")
            return _format_result(
                "search_skills",
                "success",
                results=results,
                category=category
            )

        except Exception as e:
            logger.error(f"Error in search_skills: {e}")
            return _format_result("search_skills", "error", error=str(e))

    # ========================================================================
    # TOOL 8: Search Awards and Certifications
    # ========================================================================
    def search_awards_certifications(self, award_type: Optional[str] = None) -> Dict[str, Any]:
        """
        Find awards and certifications records.

        Args:
            award_type: Type of award/certification to filter (optional)

        Returns:
            Dict with awards and certifications records
        """
        try:
            cv_id = self._get_cv_id()

            if award_type:
                results = self.db.execute_query(
                    """
                    SELECT title, issuing_organization, organization, issue_date, keywords
                    FROM awards_certifications
                    WHERE cv_id = %s AND (issuing_organization ILIKE %s OR organization ILIKE %s OR title ILIKE %s)
                    ORDER BY issue_date DESC
                    """,
                    (cv_id, f"%{award_type}%", f"%{award_type}%", f"%{award_type}%")
                )
                search_type = f"type: {award_type}"

            else:
                results = self.db.execute_query(
                    """
                    SELECT title, issuing_organization, organization, issue_date, keywords
                    FROM awards_certifications
                    WHERE cv_id = %s
                    ORDER BY issue_date DESC
                    """,
                    (cv_id,)
                )
                search_type = "all awards and certifications"

            for result in results:
                convert_dates(result, ["issue_date"])

            logger.info(f"Found {len(results)} awards/certifications for {search_type}")
            return _format_result(
                "search_awards_certifications",
                "success",
                results=results,
                search_type=search_type
            )

        except Exception as e:
            logger.error(f"Error in search_awards_certifications: {e}")
            return _format_result("search_awards_certifications", "error", error=str(e))

    # ========================================================================
    # TOOL 9: Semantic Search
    # ========================================================================
    def semantic_search(self, query: str, section: Optional[str] = None, top_k: int = 5) -> Dict[str, Any]:
        """
        Perform semantic search on CV content using vector embeddings.

        Args:
            query: Natural language search query
            section: Filter by section (work_experience, education, publication, all)
            top_k: Number of results to return (default: 5)

        Returns:
            Dict with semantic search results from Qdrant
        """
        try:
            query_embedding = self.embedding_model.embed_query(query)
            collection_name = self.config.get_qdrant_collection()

            # Build filter if section is specified
            query_filter = None
            if section and section != "all":
                query_filter = Filter(
                    must=[
                        FieldCondition(
                            key="section",
                            match=MatchValue(value=section)
                        )
                    ]
                )

            # Use the Qdrant query_points API (search is deprecated)
            response = self.qdrant_client.query_points(
                collection_name=collection_name,
                query=query_embedding,
                query_filter=query_filter,
                limit=top_k,
                with_payload=True
            )

            # query_points returns a QueryResponse, extract the points
            results = response.points

            formatted_results = []
            for result in results:
                # Build result with core fields
                formatted_result = {
                    "chunk_id": result.payload.get("chunk_id"),
                    "cv_id": result.payload.get("cv_id"),
                    "section": result.payload.get("section"),
                    "similarity_score": result.score
                }

                # Add section-specific metadata fields
                section_value = result.payload.get("section", "")

                # Work experience specific fields
                if section_value == "work experience":
                    if result.payload.get("company"):
                        formatted_result["company"] = result.payload.get("company")
                    if result.payload.get("role"):
                        formatted_result["role"] = result.payload.get("role")
                    if result.payload.get("domain"):
                        formatted_result["domain"] = result.payload.get("domain")
                    if result.payload.get("responsibility"):
                        formatted_result["responsibility"] = result.payload.get("responsibility")

                # Education specific fields
                elif section_value == "education":
                    if result.payload.get("institution"):
                        formatted_result["institution"] = result.payload.get("institution")
                    if result.payload.get("degree"):
                        formatted_result["degree"] = result.payload.get("degree")
                    if result.payload.get("thesis"):
                        formatted_result["thesis"] = result.payload.get("thesis")
                    if result.payload.get("graduation_date"):
                        formatted_result["graduation_date"] = result.payload.get("graduation_date")

                # Publication specific fields
                elif section_value == "publication":
                    if result.payload.get("title"):
                        formatted_result["title"] = result.payload.get("title")

                # Projects specific fields
                elif section_value == "projects":
                    if result.payload.get("project_name"):
                        formatted_result["project_name"] = result.payload.get("project_name")
                    if result.payload.get("responsibility"):
                        formatted_result["responsibility"] = result.payload.get("responsibility")
                    if result.payload.get("technologies"):
                        formatted_result["technologies"] = result.payload.get("technologies")

                # Common optional fields
                if result.payload.get("technologies"):
                    formatted_result["technologies"] = result.payload.get("technologies")
                if result.payload.get("skills"):
                    formatted_result["skills"] = result.payload.get("skills")

                formatted_results.append(formatted_result)

            logger.info(f"Semantic search found {len(formatted_results)} results for query: '{query}'")
            return {
                "status": "success",
                "tool": "semantic_search",
                "query": query,
                "section_filter": section or "all",
                "results_count": len(formatted_results),
                "results": formatted_results
            }

        except Exception as e:
            logger.error(f"Error in semantic_search: {e}")
            return {
                "status": "error",
                "tool": "semantic_search",
                "error": str(e)
            }


# ============================================================================
# MCP SERVER SETUP
# ============================================================================

def create_mcp_server():
    """Create and return configured MCP server"""
    try:
        config = get_config()
        tools = DatabaseTools(config)
        logger.info("MCP Server initialized successfully")
        return tools
    except Exception as e:
        logger.error(f"Failed to initialize MCP Server: {e}")
        raise


if __name__ == "__main__":
    logger.info("MCP Server started")
    try:
        mcp_server = create_mcp_server()
        logger.info("All tools available and ready")

        # Simple test: get CV summary
        result = mcp_server.get_cv_summary()
        if result["status"] == "success":
            print("✓ MCP Server is working correctly")
        else:
            print(f"✗ MCP Server error: {result.get('error')}")
    except Exception as e:
        print(f"✗ MCP Server initialization failed: {e}")

File: src/mcp_server.ts
import OpenAI from 'openai';
import type { ConfigManager } from './config.js';
import { getConfig } from './config.js';
import { getPostgresManager, getQdrantManager, type PostgreSQLManager, type QdrantManager } from './db_manager.js';
import { CVNotFoundError } from './exceptions.js';

type ToolResult = Record<string, unknown>;

// ============================================================================
// DATABASE TOOLS
// ============================================================================

export class DatabaseTools {
  private config: ConfigManager;
  private pg: PostgreSQLManager;
  private qdrant: QdrantManager;
  private openai: OpenAI;
  private _cvId: string | null = null;

  constructor(config?: ConfigManager) {
    this.config = config ?? getConfig();
    this.pg = getPostgresManager(this.config);
    this.qdrant = getQdrantManager(this.config);
    this.openai = new OpenAI({ apiKey: this.config.getOpenaiApiKey() });
  }

  async getCvId(): Promise<string> {
    if (this._cvId === null) {
      const result = await this.pg.fetchOne('SELECT id FROM cv_metadata LIMIT 1');
      if (!result) {
        throw new CVNotFoundError('No CV data found in database. Please run db_ingestion to load data.');
      }
      this._cvId = String(result['id']);
    }
    return this._cvId;
  }

  // Tool 1: Get CV Summary
  async getCvSummary(): Promise<ToolResult> {
    try {
      const result = await this.pg.fetchOne(`
        SELECT name, crole as current_role, total_years_experience,
               total_jobs, total_degrees, total_publications,
               domains, all_skills
        FROM cv_summary
        LIMIT 1
      `);
      if (result) {
        return { status: 'success', tool: 'get_cv_summary', summary: result };
      }
      return { status: 'error', tool: 'get_cv_summary', error: 'CV not found' };
    } catch (e) {
      return { status: 'error', tool: 'get_cv_summary', error: String(e) };
    }
  }

  // Tool 2: Search Company Experience
  async searchCompanyExperience(companyName: string): Promise<ToolResult> {
    try {
      const cvId = await this.getCvId();
      const results = await this.pg.fetchAll(`
        SELECT company, role, location, start_date, end_date, is_current,
               technologies, skills, domain, seniority, team_size, content
        FROM work_experience
        WHERE cv_id = $1 AND company ILIKE $2
        ORDER BY start_date DESC
      `, [cvId, `%${companyName}%`]);

      for (const r of results) {
        for (const key of ['start_date', 'end_date']) {
          if (r[key]) r[key] = String(r[key]);
        }
      }
      return { status: 'success', tool: 'search_company_experience', company: companyName, results_count: results.length, results };
    } catch (e) {
      return { status: 'error', tool: 'search_company_experience', error: String(e) };
    }
  }

  // Tool 3: Search Technology Experience
  async searchTechnologyExperience(technology: string): Promise<ToolResult> {
    try {
      const cvId = await this.getCvId();
      const results = await this.pg.fetchAll(`
        SELECT company, role, start_date, end_date, technologies, domain
        FROM work_experience
        WHERE cv_id = $1 AND $2 = ANY(technologies)
        ORDER BY start_date DESC
      `, [cvId, technology]);

      for (const r of results) {
        for (const key of ['start_date', 'end_date']) {
          if (r[key]) r[key] = String(r[key]);
        }
      }
      return { status: 'success', tool: 'search_technology_experience', technology, results_count: results.length, results };
    } catch (e) {
      return { status: 'error', tool: 'search_technology_experience', error: String(e) };
    }
  }

  // Tool 4: Search Work by Date Range
  async searchWorkByDate(startYear: number, endYear: number): Promise<ToolResult> {
    try {
      const cvId = await this.getCvId();
      const results = await this.pg.fetchAll(`
        SELECT company, role, start_date, end_date, technologies, keywords
        FROM work_experience
        WHERE cv_id = $1
          AND start_date >= $2::date
          AND (end_date <= $3::date OR end_date IS NULL)
        ORDER BY start_date DESC
      `, [cvId, `${startYear}-01-01`, `${endYear}-12-31`]);

      for (const r of results) {
        for (const key of ['start_date', 'end_date']) {
          if (r[key]) r[key] = String(r[key]);
        }
      }
      return { status: 'success', tool: 'search_work_by_date', date_range: `${startYear}-${endYear}`, results_count: results.length, results };
    } catch (e) {
      return { status: 'error', tool: 'search_work_by_date', error: String(e) };
    }
  }

  // Tool 5: Search Education
  async searchEducation(institution?: string, degree?: string): Promise<ToolResult> {
    try {
      const cvId = await this.getCvId();
      let results: Record<string, unknown>[];
      let searchType: string;

      if (institution) {
        results = await this.pg.fetchAll(`
          SELECT institution, degree, field, specialization, graduation_date, thesis, publications, content
          FROM education
          WHERE cv_id = $1 AND institution ILIKE $2
        `, [cvId, `%${institution}%`]);
        searchType = `institution: ${institution}`;
      } else if (degree) {
        results = await this.pg.fetchAll(`
          SELECT institution, degree, field, specialization, graduation_date, thesis, content
          FROM education
          WHERE cv_id = $1 AND degree ILIKE $2
        `, [cvId, `%${degree}%`]);
        searchType = `degree: ${degree}`;
      } else {
        results = await this.pg.fetchAll(`
          SELECT institution, degree, field, specialization, graduation_date, thesis, content
          FROM education
          WHERE cv_id = $1
        `, [cvId]);
        searchType = 'all education';
      }

      for (const r of results) {
        if (r['graduation_date']) r['graduation_date'] = String(r['graduation_date']);
      }
      return { status: 'success', tool: 'search_education', search_type: searchType, results_count: results.length, results };
    } catch (e) {
      return { status: 'error', tool: 'search_education', error: String(e) };
    }
  }

  // Tool 6: Search Publications
  async searchPublications(year?: number): Promise<ToolResult> {
    try {
      const cvId = await this.getCvId();
      let results: Record<string, unknown>[];
      let searchType: string;

      if (year) {
        results = await this.pg.fetchAll(`
          SELECT title, year, conference_name, doi, keywords, content_text
          FROM publications
          WHERE cv_id = $1 AND year = $2
          ORDER BY year DESC
        `, [cvId, year]);
        searchType = `year: ${year}`;
      } else {
        results = await this.pg.fetchAll(`
          SELECT title, year, conference_name, doi, keywords, content_text
          FROM publications
          WHERE cv_id = $1
          ORDER BY year DESC
        `, [cvId]);
        searchType = 'all publications';
      }
      return { status: 'success', tool: 'search_publications', search_type: searchType, results_count: results.length, results };
    } catch (e) {
      return { status: 'error', tool: 'search_publications', error: String(e) };
    }
  }

  // Tool 7: Search Skills
  async searchSkills(category: string): Promise<ToolResult> {
    try {
      const cvId = await this.getCvId();
      const results = await this.pg.fetchAll(`
        SELECT skill_name
        FROM skills
        WHERE cv_id = $1 AND skill_category = $2
        ORDER BY skill_name
      `, [cvId, category]);
      return { status: 'success', tool: 'search_skills', category, results_count: results.length, results };
    } catch (e) {
      return { status: 'error', tool: 'search_skills', error: String(e) };
    }
  }

  // Tool 8: Search Awards and Certifications
  async searchAwardsCertifications(awardType?: string): Promise<ToolResult> {
    try {
      const cvId = await this.getCvId();
      let results: Record<string, unknown>[];
      let searchType: string;

      if (awardType) {
        results = await this.pg.fetchAll(`
          SELECT title, issuing_organization, organization, issue_date, keywords, content
          FROM awards_certifications
          WHERE cv_id = $1 AND (issuing_organization ILIKE $2 OR organization ILIKE $2 OR title ILIKE $2)
          ORDER BY issue_date DESC
        `, [cvId, `%${awardType}%`]);
        searchType = `type: ${awardType}`;
      } else {
        results = await this.pg.fetchAll(`
          SELECT title, issuing_organization, organization, issue_date, keywords, content
          FROM awards_certifications
          WHERE cv_id = $1
          ORDER BY issue_date DESC
        `, [cvId]);
        searchType = 'all awards and certifications';
      }

      for (const r of results) {
        if (r['issue_date']) r['issue_date'] = String(r['issue_date']);
      }
      return { status: 'success', tool: 'search_awards_certifications', search_type: searchType, results_count: results.length, results };
    } catch (e) {
      return { status: 'error', tool: 'search_awards_certifications', error: String(e) };
    }
  }

  // Tool 9: Semantic Search
  async semanticSearch(query: string, section?: string, topK: number = 5): Promise<ToolResult> {
    try {
      const embeddingResponse = await this.openai.embeddings.create({
        model: 'text-embedding-3-small',
        input: query,
      });
      const vector = embeddingResponse.data[0].embedding;

      const searchParams: Record<string, unknown> = { limit: topK, with_payload: true };
      if (section && section !== 'all') {
        searchParams['filter'] = {
          must: [{ key: 'section', match: { value: section } }],
        };
      }

      const results = await this.qdrant.client.search(
        this.config.getQdrantCollection(),
        { vector, ...searchParams } as Parameters<typeof this.qdrant.client.search>[1],
      );

      const formattedResults = results.map((result) => {
        const payload = result.payload ?? {};
        const sectionValue = String(payload['section'] ?? '');
        const fr: Record<string, unknown> = {
          chunk_id: payload['chunk_id'],
          cv_id: payload['cv_id'],
          section: sectionValue,
          similarity_score: result.score,
        };

        if (sectionValue === 'work experience') {
          if (payload['company']) fr['company'] = payload['company'];
          if (payload['role']) fr['role'] = payload['role'];
          if (payload['domain']) fr['domain'] = payload['domain'];
          if (payload['responsibility']) fr['responsibility'] = payload['responsibility'];
        } else if (sectionValue === 'education') {
          if (payload['institution']) fr['institution'] = payload['institution'];
          if (payload['degree']) fr['degree'] = payload['degree'];
          if (payload['thesis']) fr['thesis'] = payload['thesis'];
          if (payload['graduation_date']) fr['graduation_date'] = payload['graduation_date'];
          if (payload['description']) fr['description'] = payload['description'];
        } else if (sectionValue === 'publication') {
          if (payload['title']) fr['title'] = payload['title'];
          if (payload['description']) fr['description'] = payload['description'];
        } else if (sectionValue === 'projects') {
          if (payload['project_name']) fr['project_name'] = payload['project_name'];
          if (payload['responsibility']) fr['responsibility'] = payload['responsibility'];
          if (payload['technologies']) fr['technologies'] = payload['technologies'];
          if (payload['description']) fr['description'] = payload['description'];
        }

        if (payload['technologies']) fr['technologies'] = payload['technologies'];
        if (payload['skills']) fr['skills'] = payload['skills'];
        if (payload['description'] && !fr['description']) fr['description'] = payload['description'];

        return fr;
      });

      return { status: 'success', tool: 'semantic_search', query, section_filter: section ?? 'all', results_count: formattedResults.length, results: formattedResults };
    } catch (e) {
      return { status: 'error', tool: 'semantic_search', error: String(e) };
    }
  }

  // Tool 10: Get All Work Experience
  async getAllWorkExperience(): Promise<ToolResult> {
    try {
      const cvId = await this.getCvId();
      const results = await this.pg.fetchAll(`
        SELECT company, role, location, start_date, end_date, is_current,
               technologies, skills, domain, seniority, team_size, content
        FROM work_experience
        WHERE cv_id = $1
        ORDER BY start_date DESC
      `, [cvId]);

      for (const r of results) {
        for (const key of ['start_date', 'end_date']) {
          if (r[key]) r[key] = String(r[key]);
        }
      }
      return { status: 'success', tool: 'get_all_work_experience', results_count: results.length, results };
    } catch (e) {
      return { status: 'error', tool: 'get_all_work_experience', error: String(e) };
    }
  }

  // Tool 11: Search Languages
  async searchLanguages(language?: string): Promise<ToolResult> {
    try {
      const cvId = await this.getCvId();
      let results: Record<string, unknown>[];
      let searchType: string;

      if (language) {
        results = await this.pg.fetchAll(`
          SELECT language, proficiency_level
          FROM languages
          WHERE cv_id = $1 AND language ILIKE $2
          ORDER BY language
        `, [cvId, `%${language}%`]);
        searchType = `language: ${language}`;
      } else {
        results = await this.pg.fetchAll(`
          SELECT language, proficiency_level
          FROM languages
          WHERE cv_id = $1
          ORDER BY language
        `, [cvId]);
        searchType = 'all languages';
      }
      return { status: 'success', tool: 'search_languages', search_type: searchType, results_count: results.length, results };
    } catch (e) {
      return { status: 'error', tool: 'search_languages', error: String(e) };
    }
  }

  // Tool 12: Get Contact Info
  async getContactInfo(): Promise<ToolResult> {
    try {
      const cvId = await this.getCvId();
      const result = await this.pg.fetchOne(`
        SELECT name, email, email_alt, linkedin, github
        FROM cv_metadata
        WHERE id = $1
      `, [cvId]);
      if (result) {
        return { status: 'success', tool: 'get_contact_info', data: result };
      }
      return { status: 'error', tool: 'get_contact_info', error: 'Contact information not found' };
    } catch (e) {
      return { status: 'error', tool: 'get_contact_info', error: String(e) };
    }
  }

  // Tool 13: Search Work References
  async searchWorkReferences(referenceName?: string, company?: string): Promise<ToolResult> {
    try {
      const cvId = await this.getCvId();
      let results: Record<string, unknown>[];
      let searchType: string;

      if (referenceName) {
        results = await this.pg.fetchAll(`
          SELECT name, position, company, email, note
          FROM work_references
          WHERE cv_id = $1 AND name ILIKE $2
          ORDER BY name
        `, [cvId, `%${referenceName}%`]);
        searchType = `name: ${referenceName}`;
      } else if (company) {
        results = await this.pg.fetchAll(`
          SELECT name, position, company, email, note
          FROM work_references
          WHERE cv_id = $1 AND company ILIKE $2
          ORDER BY name
        `, [cvId, `%${company}%`]);
        searchType = `company: ${company}`;
      } else {
        results = await this.pg.fetchAll(`
          SELECT name, position, company, email, note
          FROM work_references
          WHERE cv_id = $1
          ORDER BY name
        `, [cvId]);
        searchType = 'all references';
      }
      return { status: 'success', tool: 'search_work_references', search_type: searchType, results_count: results.length, results };
    } catch (e) {
      return { status: 'error', tool: 'search_work_references', error: String(e) };
    }
  }
}
